RedditSharp/WebAgent.cs

```diff
@@ -112,13 +66,38 @@ public int RequestsThisBurst
         public WebAgent( string accessToken ) {
             RootDomain = OAuthDomainUrl;
             AccessToken = accessToken;
+            RateLimitUsed = 0;
+            RateLimitReset = DateTimeOffset.UtcNow;
+            RateLimitRemaining = IsOAuth ? 60 : 30;
         }
 
         /// <inheritdoc />
-        public virtual async Task<JToken> ExecuteRequestAsync(HttpRequestMessage request)
+        public virtual async Task<JToken> ExecuteRequestAsync(Func<HttpRequestMessage> request)
         {
-            await EnforceRateLimit().ConfigureAwait(false);
-            var response = await _httpClient.SendAsync(request).ConfigureAwait(false);
+            if (request == null)
+              throw new ArgumentNullException(nameof(request));
+            const int maxTries = 20;
+            HttpResponseMessage response;
+            var tries = 0;
+            do {
+              await rateLimitLock.WaitAsync().ConfigureAwait(false);
```

Comment:
This may result in race conditions and potential violations of the rate limit with enough parallelism, even before it gets responses.

The header in the latest arriving response may not necessarily be the latest processed result. We'll need to track when each request was sent as well as X-Ratelimit-Reset.